File: RobustUnitTestingInAndroid/src/test/java/com/code/sliski/mainscreen/ui/MainActivityPresenterTest.java
package com.code.sliski.mainscreen.ui;

import android.os.Bundle;
import com.code.sliski.UserRole;
import org.junit.Test;
import org.junit.experimental.runners.Enclosed;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;
import org.junit.runners.Parameterized.Parameters;

import java.util.Arrays;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

import static org.mockito.Mockito.*;

@RunWith(Enclosed.class)
public class MainActivityPresenterTest {

    static Map<UserRole, String> userRoleDictionary = new HashMap<>();

    static {
        userRoleDictionary.put(UserRole.HERO, "You are a Hero.");
        userRoleDictionary.put(UserRole.GRUNT, "You are a Grunt.");
        userRoleDictionary.put(UserRole.PEON, "You are a Peon.");
    }

    public static class IndividualTests {

        private MainActivityPresenter mainActivityPresenter;

        // Mocks
        private View mainActivityPresenterView;

        @Test
        public void presentTestShouldNotInteractWithView() throws Exception {
            // GIVEN
            mainActivityPresenterView = mock(View.class);
            mainActivityPresenter = new MainActivityPresenter(1L, UserRole.HERO, userRoleDictionary);
            mainActivityPresenter.attach(mainActivityPresenterView);

            // WHEN
            mainActivityPresenter.present(mock(Bundle.class));

            // THEN
            verifyZeroInteractions(mainActivityPresenterView);
        }

        @Test
        public void presentTestShouldShowLoginFragment() throws Exception {
            // GIVEN
            mainActivityPresenterView = mock(View.class);
            mainActivityPresenter = new MainActivityPresenter(0L, UserRole.HERO, userRoleDictionary);
            mainActivityPresenter.attach(mainActivityPresenterView);

            // WHEN
            mainActivityPresenter.present(null);

            // THEN
            verify(mainActivityPresenterView, times(1)).showLoginScreen();
        }

        @Test
        public void presentTestShouldShowUserInfoScreen() throws Exception {
            // GIVEN
            mainActivityPresenterView = mock(View.class);
            mainActivityPresenter = new MainActivityPresenter(1L, UserRole.HERO, userRoleDictionary);
            mainActivityPresenter.attach(mainActivityPresenterView);

            // WHEN
            mainActivityPresenter.present(null);

            // THEN
            verify(mainActivityPresenterView, times(1)).showUserInfoScreen();
        }
    }

    @RunWith(Parameterized.class)
    public static class ParameterizedTests {

        private MainActivityPresenter mainActivityPresenter;

        // Mocks
        private View mainActivityPresenterView;

        @Parameters(name = "{2}")
        public static Collection<Object[]> data() {
            return Arrays.asList(new Object[][]{
                    {1, UserRole.HERO, "You are a Hero.", 1},
                    {1, UserRole.GRUNT, "You are a Grunt.", 1},
                    {1, UserRole.PEON, "You are a Peon.", 1},
                    {0, UserRole.HERO, "No Message", 0},
                    {0, UserRole.GRUNT, "No Message", 0},
                    {0, UserRole.PEON, "No Message", 0}
            });
        }

        private long userId;
        private UserRole userRole;
        private String message;
        private int times;

        public ParameterizedTests(long userId, UserRole userRole, String message, int times) {
            this.userId = userId;
            this.userRole = userRole;
            this.message = message;
            this.times = times;
        }

        @Test
        public void presentTestShouldDisplay() {
            // GIVEN
            mainActivityPresenterView = mock(View.class);
            mainActivityPresenter = new MainActivityPresenter(userId, userRole, userRoleDictionary);
            mainActivityPresenter.attach(mainActivityPresenterView);

            // WHEN
            mainActivityPresenter.present(null);

            // THEN
            verify(mainActivityPresenterView, times(times)).displayMessage(message);
        }
    }
}

File: RobustUnitTestingInAndroid/src/test/java/com/code/sliski/mainscreen/ui/MainActivityPresenterTest.kt
package com.code.sliski.mainscreen.ui

import android.os.Bundle
import com.code.sliski.UserRole
import org.junit.Test
import org.junit.experimental.runners.Enclosed
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import org.junit.runners.Parameterized.Parameters
import org.mockito.Mockito.*

var userRoleDictionary: Map<UserRole, String> = mapOf(Pair(UserRole.HERO, "You are a Hero."),
                                                      Pair(UserRole.GRUNT, "You are a Grunt."),
                                                      Pair(UserRole.PEON, "You are a Peon."))

@RunWith(Enclosed::class)
class MainActivityPresenterTest {

    class IndividualTests {

        lateinit var tested: MainActivityPresenter
        val viewMock = mock(View::class.java)

        fun setupPresenter(userId: Long) {
            tested = MainActivityPresenter(userId, UserRole.HERO, userRoleDictionary)
            tested.attach(viewMock)
        }

        @Test @Throws(Exception::class)
        fun `not interact with view`() {
            setupPresenter(1L)

            tested.present(state = mock(Bundle::class.java))

            verifyZeroInteractions(viewMock)
        }

        @Test @Throws(Exception::class)
        fun `show login fragment`() {
            setupPresenter(0L)

            tested.present(state = null)

            verify(viewMock, times(1)).showLoginScreen()
        }

        @Test @Throws(Exception::class)
        fun `not show login fragment`() {
            setupPresenter(1L)

            tested.present(state = null)

            verify(viewMock, never()).showLoginScreen()
        }

        @Test @Throws(Exception::class)
        fun `show user info screen`() {
            setupPresenter(1L)

            tested.present(state = null)

            verify(viewMock, times(1)).showUserInfoScreen()
        }

        @Test @Throws(Exception::class)
        fun `not show user info screen`() {
            setupPresenter(0L)

            tested.present(state = null)

            verify(viewMock, never()).showUserInfoScreen()
        }
    }

    @RunWith(Parameterized::class)
    class ParameterizedTests(var userId: Long,
                             var userRole: UserRole,
                             var message: String,
                             var times: Int) {

        lateinit var tested: MainActivityPresenter
        val mockView = mock(View::class.java)

        companion object {
            @Parameters(name = "{2}")
            @JvmStatic fun data() = listOf(
                    arrayOf(1, UserRole.HERO, "You are a Hero.", 1),
                    arrayOf(1, UserRole.GRUNT, "You are a Grunt.", 1),
                    arrayOf(1, UserRole.PEON, "You are a Peon.", 1),
                    arrayOf(0, UserRole.HERO, "No Message", 0),
                    arrayOf(0, UserRole.GRUNT, "No Message", 0),
                    arrayOf(0, UserRole.PEON, "No Message", 0))

        }

        @Test @Throws(Exception::class)
        fun `display message`() {
            tested = MainActivityPresenter(userId, userRole, userRoleDictionary)
            tested.attach(mockView)

            tested.present(state = null)

            verify(mockView, times(times)).displayMessage(message)
        }
    }
}